akka-cluster-sharding/src/main/scala/akka/cluster/sharding/WowAllocationStrategy.scala

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (C) 2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.cluster.sharding
+
+import scala.collection.immutable
+import scala.concurrent.Future
+
+import akka.actor.ActorRef
+import akka.cluster.sharding.ShardCoordinator.ShardAllocationStrategy
+import akka.cluster.sharding.ShardRegion.ShardId
+
+object WowAllocationStrategy {
+  private val emptyRebalanceResult = Future.successful(Set.empty[ShardId])
+}
+
+// FIXME name
+class WowAllocationStrategy(absoluteLimit: Int, relativeLimit: Double) extends ShardAllocationStrategy {
+  import WowAllocationStrategy.emptyRebalanceResult
+
+  override def allocateShard(
+      requester: ActorRef,
+      shardId: ShardId,
+      currentShardAllocations: Map[ActorRef, immutable.IndexedSeq[ShardId]]): Future[ActorRef] = {
+    val (regionWithLeastShards, _) = currentShardAllocations.minBy { case (_, v) => v.size }
+    Future.successful(regionWithLeastShards)
+  }
+
+  override def rebalance(
+      currentShardAllocations: Map[ActorRef, immutable.IndexedSeq[ShardId]],
+      rebalanceInProgress: Set[ShardId]): Future[Set[ShardId]] = {
+    import math.{ max, min }
+
+    if (rebalanceInProgress.nonEmpty) {
+      // one rebalance at a time
+      emptyRebalanceResult
+    } else {
+      val numberOfShards = currentShardAllocations.valuesIterator.map(_.size).sum
+      val numberOfRegions = currentShardAllocations.size
+      val optimalPerRegion = numberOfShards / numberOfRegions + (if (numberOfShards % numberOfRegions == 0) 0 else 1)
+
+      def limit(resultSize: Int): Int =
+        max(1, min(min((relativeLimit * numberOfShards).toInt, resultSize), min(absoluteLimit, resultSize)))
+
+      val selected = Vector.newBuilder[ShardId]
+      val sortedAllocations = currentShardAllocations.valuesIterator.toVector.sortBy(_.size)
+      sortedAllocations.foreach { shards =>
+        if (shards.size > optimalPerRegion) {
+          selected ++= shards.take(shards.size - optimalPerRegion)
+        }
+      }
+      val result = selected.result()
+      val limitedResult = result.take(limit(result.size)).toSet
+
+      if (limitedResult.nonEmpty) {
+        println(
+          s"# rebalance, currentShardAllocations [${currentShardAllocations.valuesIterator.map(_.size).mkString(",")}], " +
+          s"numberOfShards $numberOfShards, numberOfRegions, $numberOfRegions, optimalPerRegion $optimalPerRegion, " +
+          s"currentShards [${currentShardAllocations.valuesIterator.map(_.sorted.mkString("(", ", ", ")")).mkString(" ")}] " +
+          s"limit ${limit(result.size)}, result [${limitedResult.toList.sorted.mkString(",")}]") // FIXME
+
+        Future.successful(limitedResult)
+      } else {
+        // in second phase we look for diff of 2
```

Comment:
In the first phase the optimal is rounded up, and depending on number of shards per region and number of regions that might not be the exact optimal. Example from the test: `[5, 5, 0]`. optimal from first phase is 4 but that results in `[4, 4, 2]`. In second phase it can find that diff of 2 (4-2) and move one more shard => `[3, 4, 3]`.

For more complex setups it can be larger diffs, and I have updated the PR to handle that better now.